samples/drivers/gecko_adc/src/bq24250.c

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright (c) 2022 T-Mobile USA, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/byteorder.h>
+#include "bq24250.h"
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(bq24250), okay)
+
+#define BQ_GPIO_F_NAME               "GPIO_F"
+#define BQ_CHARGE_INT                12      /* PF12 - configured here */
+#define BQ_CHARGE_CEn                13      /* PF13 - set low in board.c */
+#define BQ_ADDRESS                   0x6a
+
+#define ENABLE_HARDWARE              90
+#define BQ_INIT                      91
+#define BQ_INT_MONITOR               92
+
+#define MAX_BYTES_FOR_REGISTER_INDEX 4
+#define MAX_I2C_BYTES                20
+
+#define BQ_CHARGER_FAULT_MODE        3
+#define BQ_SLEEP_MODE                3
+#define INPUT_FAULT_AND_LDO_LOW_MODE 10
+
+struct drv_data {
+	struct gpio_callback gpio_int_cb;
+	gpio_flags_t mode;
+	int index;
+	int aux;
+};
+
+bool restart_setup;
+
+/* BQ Battery and Charger Status */
+const struct device *gpiof_dev = DEVICE_DT_GET(DT_NODELABEL(gpiof));
+
+static const struct device *i2c1_dev = DEVICE_DT_GET(DT_NODELABEL(i2c1));
+
+/* BQ Battery Charger interrupt callback */
+static struct gpio_callback gpio_batt_chrg_cb;
+int bq_battery_charge_int_isr_count;
+bool bq_battery_charge_int_state_change;
+uint8_t i2c_read_data[20];
+
+static int i2c_read_to_buffer(uint8_t dev_addr, int reg_addr, uint8_t *buf, uint8_t buf_length)
+{
+	const struct device *dev;
+	uint8_t reg_addr_buf[MAX_BYTES_FOR_REGISTER_INDEX];
+	int reg_addr_bytes;
+	int ret;
+
+	dev = i2c1_dev;
+	if (!dev) {
+		printf("I2C: Device driver I2C_1 not found.");
+		return -ENODEV;
+	}
+
+	reg_addr_bytes = 1;
+	sys_put_be32(reg_addr, reg_addr_buf);
+
+	ret = i2c_write_read(dev, dev_addr,
+			reg_addr_buf +
+			MAX_BYTES_FOR_REGISTER_INDEX - reg_addr_bytes,
+			reg_addr_bytes, buf, buf_length);
+	if (ret < 0) {
+		printf("Failed to read from device: 0x%x", dev_addr);
+		return -EIO;
+	}
+	return 0;
+}
+
+int get_bq24250_status(uint8_t *charging, uint8_t *vbus, uint8_t *attached, uint8_t *fault)
+{
+	int ret = 0;
+	uint8_t bq_addr = 0x6a;
+	uint8_t rx_len = 1;
+	uint8_t bq_reg_0x00 = 0x00;
+	uint8_t bq_reg_0x00_data;
+
+	ret = i2c_read_to_buffer(bq_addr, bq_reg_0x00, &bq_reg_0x00_data, rx_len);
+	if (ret) {
+		printf("read_bytes function:Error reading from BQ! error code (%d)\n", ret);
+	}
```

Comment:
If the i2c_read_to_buffer() call fails, an "else" statement could be used here to avoid subsequent processing before returning from this function.  Of course, the storage to which the "vbus," "charging," "attached," and "fault" pointers reference must then be set accordingly.

Reply:
From my understanding of the code (written by someone else), it looks like the section of code you are talking about needs to be processed in both error and normal conditions.  An else statement would only process that block of code in a normal condition.

Reply:
The data conveyed to the caller(s) needs to be appropriately processed irrespective of the return status of i2c_read_to_buffer().